man/systeminfo.8
.\" Automatically generated by Pod::Man 2.27 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "SYSTEMINFO_POD 1"
.TH SYSTEMINFO_POD 1 "2015-10-04" "perl v5.18.2" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
systeminfo \- utility for displaying hardware information
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
systeminfo \-\-get asset_type [\-\-p|\-\-l|\-\-j|\-\-d identifier] [\-\-c]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
systeminfo is utility for getting hardware information
it aims to be simple and provide output in well formatted
output
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fBasset_type\fR" 8
.IX Item "asset_type"
can be one of these types: system, cpu, memory, disk, pci, fcms, tape, eth
.IP "\fB\-\-l, \-\-long\fR" 8
.IX Item "--l, --long"
specifies to display long output
.IP "\fB\-\-p, \-\-parsable\fR" 8
.IX Item "--p, --parsable"
specifies to display parsable output
.IP "\fB\-\-j, \-\-json\fR" 8
.IX Item "--j, --json"
specifies to display output in json
.IP "\fB\-\-d, \-\-detail \f(BIidentifier\fB \fR" 8
.IX Item "--d, --detail identifier "
specifies to display detail, requires identifier
.Sp
identifier
.Sp
column which you should use as identifier is marked 
in column header with asterisk
.IP "\fB\-\-c, \-\-cached\fR" 8
.IX Item "--c, --cached"
use cache to get data, should be faster, but doesn't generate fresh data
.SH "EXAMPLES"
.IX Header "EXAMPLES"
This gets information about system in short format:
.PP
.Vb 1
\&    systeminfo \-\-get system
.Ve
.PP
This gets information about disks in long format:
.PP
.Vb 3
\&    systeminfo \-\-get disk \-\-l
\&    or
\&    systeminfo \-\-get disk \-\-long
.Ve
.PP
This gets information about fcms \s-1HBA\s0's in parsable format:
.PP
.Vb 1
\&    systeminfo \-\-get fcms \-\-p
.Ve
.PP
This gets information about memory in json format:
.PP
.Vb 1
\&    systeminfo \-\-get memory \-\-j
.Ve
.PP
This get detail about disk device:
.PP
.Vb 1
\&    systeminfo \-\-get disk \-\-detail 24:0:2:0
.Ve
.PP
This refreshes cache info about disks:
.PP
.Vb 3
\&    systeminfo \-\-get disk
\&    systeminfo \-\-get disk \-\-l
\&    systeminfo \-\-get disk \-\-p
.Ve
.PP
This doesn't refresh cache:
.PP
.Vb 8
\&    (gets fresh data but doesn\*(Aqt update cache)
\&    systeminfo \-\-get disk \-\-detail 24:0:2:0
\&    or (these two examples get data from cache)
\&    systeminfo \-\-get disk \-\-detail 24:0:2:0 \-\-c
\&    or
\&    systeminfo \-\-get disk \-\-c
\&    or
\&    systeminfo \-\-get disk \-\-l \-\-c
.Ve
.SH "NOTES"
.IX Header "NOTES"
This utility should be run with root priveleges
.SH "AUTHOR"
.IX Header "AUTHOR"
Pavol Ipoth
.SH "COPYRIGHT AND LICENSE"
.IX Header "COPYRIGHT AND LICENSE"
Copyright 2013, 2014, 2015 Pavol Ipoth
.PP
GPLv3


systeminfo.py
#!/usr/bin/python
#
# Systeminfo - Simple utility for gathering hardware summary information
# Copyright (C) 2013, 2014  Pavol Ipoth  <pavol.ipoth@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.
#
# -*- coding: utf-8 -*-

"""
SystemInfo

Module: systeminfo.py

This python script was written for system administration purposes, to have command line
tool where you can have good and quick overview of properties of different hardware types.

@author: Pavol Ipoth
@license: GPLv3+
@copyright: Copyright 2015 Pavol Ipoth
@contact: pavol.ipoth@gmail.com
@version: 1.5.0

"""

import os
import sys

# we are checking if effective user runnig script is root
# because of using sysfs we can get results just with root permissions
effective_uid = os.geteuid()

if effective_uid > 0:
    print "To run this utility you need root priveleges!"
    sys.exit(4)

import argparse
import getopt
import re
import systeminfo.misc.config
from string import Template
from systeminfo.proc.cpu import Cpu
from systeminfo.proc.memory import Memory
from systeminfo.proc.pci import Pci
from systeminfo.proc.fcms import Fcms
from systeminfo.proc.disk import Disk
from systeminfo.proc.system import System
from systeminfo.proc.tape import Tape
from systeminfo.proc.eth import Eth

options = {'outlength': 'short', 'get_data_action': 'getData'}
"""
@type: dict
@var: this is variable with default values passed to action method
"""

asset_types = ['cpu', 'memory', 'pci', 'fcms', 'disk', 'system', 'tape', 'eth']
"""
@type: list
@var: list of asset types currently supported by tool
"""

def main():
    """
    Function: main

    This is the main function of the tool, processes options and invokes
    responsible classes and methods

    @rtype: void
    """

    # Action which we want to perform on asset_types
    action = ''

    # Asset type on which we want to perform aciont
    asset_param = ''

    parser = argparse.ArgumentParser(
        description="description: Utility for viewing HW information"
    )
    
    group = parser.add_mutually_exclusive_group()
    group.add_argument(
        "-l", 
        "--long", 
        help="display in long table format",
        action='store_const', 
        dest='outlength',
        const='long'
    )
    group.add_argument(
        "-p", 
        "--parsable",
        help="display in parsable format",
        action='store_const', 
        dest='outlength',
        const='parsable'
    )
    group.add_argument(
        "-d",
        "--detail",
        help="display detail of hw instance",
        dest='identifier'
    )
    group.add_argument(
        "-j",
        "--json",
        help="display results in json format",
        action='store_const', 
        dest='outlength',
        const='json'
    )
    parser.add_argument(
        "-g", 
        "--get", 
        help="specify type of hardware to get info for, required argument",
        choices=asset_types,
        required=True
    )
    parser.add_argument(
        "-c", 
        "--cached",
        help="display results from cache",
        action='store_const', 
        dest='',
        const='getCache'
    )

    results = parser.parse_args()
    
    if results.outlength:
        options['outlength'] = results.outlength
    
    action = 'summary'
    asset_param = results.get
    options['template_body_type'] = 'TableTemplate'
    options['template_header_type'] = 'HeaderTableTemplate'

    if results.get == 'system':
        options['template_body_type'] = 'PropertyTemplate'
        options['template_header_type'] = 'VoidTemplate'

    if results.identifier:
        action = 'detail'
        options['template_body_type'] = 'PropertyTemplate'
        options['template_header_type'] = 'VoidTemplate'
        options['outlength'] = 'detail'
        options['instance'] = results.identifier

    if results.outlength == 'json':
        action = 'summary'
        options['template_body_type'] = 'PropertyTemplateAll'
        options['template_header_type'] = 'VoidTemplate'

    # instantiating object, his class is asset type
    # and then invoking action on it, depending on passed options
    # , also passing parameters assigned in previous option checking
    asset_type = asset_param.title()
    configDir = systeminfo.misc.config.confDir
    cachingDir = systeminfo.misc.config.cacheDir

    asset = globals()[asset_type](configDir, cachingDir)
    getattr(asset, action)(options)

main()
